src/datetime.rs

```diff
@@ -0,0 +1,341 @@
+//! Support datetimes and timedeltas
+//!
+//! [The corresponding section][datetime] of the NumPy documentation contains more information.
+//!
+//! # Example
+//!
+//! ```
+//! use numpy::{datetime::{units, Datetime, Timedelta}, PyArray1};
+//! use pyo3::Python;
+//! # use pyo3::types::PyDict;
+//!
+//! Python::with_gil(|py| {
+//! #    let locals = py
+//! #        .eval("{ 'np': __import__('numpy') }", None, None)
+//! #        .unwrap()
+//! #        .downcast::<PyDict>()
+//! #        .unwrap();
+//! #
+//!     let array = py
+//!         .eval(
+//!             "np.array([np.datetime64('2017-04-21')])",
+//!             None,
+//!             Some(locals),
+//!         )
+//!         .unwrap()
+//!         .downcast::<PyArray1<Datetime<units::Days>>>()
+//!         .unwrap();
+//!
+//!     assert_eq!(
+//!         array.get_owned(0).unwrap(),
+//!         Datetime::<units::Days>::from(17_277)
+//!     );
+//!
+//!     let array = py
+//!         .eval(
+//!             "np.array([np.datetime64('2022-03-29')]) - np.array([np.datetime64('2017-04-21')])",
+//!             None,
+//!             Some(locals),
+//!         )
+//!         .unwrap()
+//!         .downcast::<PyArray1<Timedelta<units::Days>>>()
+//!         .unwrap();
+//!
+//!     assert_eq!(
+//!         array.get_owned(0).unwrap(),
+//!         Timedelta::<units::Days>::from(1_803)
+//!     );
+//! });
+//! ```
+//!
+//! [datetime]: https://numpy.org/doc/stable/reference/arrays.datetime.html
+
+use std::cell::UnsafeCell;
+use std::collections::hash_map::Entry;
+use std::fmt;
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+use ahash::AHashMap;
+use pyo3::{Py, Python};
+
+use crate::dtype::{Element, PyArrayDescr};
+use crate::npyffi::{PyArray_DatetimeDTypeMetaData, NPY_DATETIMEUNIT, NPY_TYPES};
+
+/// Represents the [datetime units][datetime-units] supported by NumPy
+///
+/// [datetime-units]: https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units
+pub trait Unit: Send + Sync + Clone + Copy + PartialEq + Eq + Hash + PartialOrd + Ord {
```

Comment:
Don't we need a private guard with this trait?

Reply:
I thought about sealing the trait but I don't really see what could go wrong if someone implements it. The could specify a useless combination of `UNIT and `ABBREV` but that seems about it, doesn't it? More specifically I don't see how it could cause memory unsafety or something like that.

Reply:
Ah OK, I think it's fine since the unit is limited to our C enum type.